package org.stok.client.protocol;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.*;

public class ProtocolParser {
    private final ObjectMapper objMapper;

    public ProtocolParser() {
        this.objMapper = new ObjectMapper();

        objMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, true);
        objMapper.configure(DeserializationFeature.USE_BIG_DECIMAL_FOR_FLOATS, true);
        objMapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
        objMapper.enable(JsonGenerator.Feature.WRITE_BIGDECIMAL_AS_PLAIN);
    }

    protected JsonNode classToNode(Object req) {
        return objMapper.valueToTree(req);
    }

    protected String nodeToJson(JsonNode node) throws JsonProcessingException {
        ObjectWriter objWritter = objMapper.writer();
//        objWritter = objWritter.with(SerializationFeature.INDENT_OUTPUT);
        return objWritter.writeValueAsString(node);
    }

    protected JsonNode jsonToNode(String req) throws JsonProcessingException {
        return objMapper.readTree(req);
    }

    protected <A> A nodeToClass(JsonNode node, Class<A> req) throws JsonProcessingException {
        return objMapper.treeToValue(node, req);
    }

    public <A> A parseResponse(String resJson, Class<A> resClass) throws JsonProcessingException {
        JsonNode resNode = jsonToNode(resJson);
        return nodeToClass(resNode, resClass);
    }

    public String parseRequest(Object reqObj) throws JsonProcessingException {
        JsonNode resNode = classToNode(reqObj);
        return nodeToJson(resNode);
    }
}
